File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.cpp
#include <string>
#include <vector>
#include <stack>

int solution(std::vector<std::vector<int>> board, std::vector<int> moves) {
    int rows = board.size();
    int cols = board[0].size();
    
    std::vector<std::stack<int>> lanes(cols);
    for (int i = rows - 1; i >= 0; --i) {
        for (int j = 0; j < cols; ++j) {
            if (board[i][j] != 0) {
                lanes[j].push(board[i][j]);
            }
        }
    }
    
    std::stack<int> bucket;
    int score = 0;
    
    for (int move : moves) {
        int lane_index = move - 1;
        if (!lanes[lane_index].empty()) {
            int doll = lanes[lane_index].top();
            lanes[lane_index].pop();
            
            if (!bucket.empty() && bucket.top() == doll) {
                bucket.pop();
                score += 2;
            } else {
                bucket.push(doll);
            }
        }
    }
    
    return score;
}

File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.java
import java.util.ArrayDeque;

class Solution {
    public int solution(int[][] board, int[] moves) {
        
        int boardSize = board.length;
        ArrayDeque<Integer>[] lanes = initializeLanes(board);
        ArrayDeque<Integer> bucket = new ArrayDeque<>();
        int removedDolls = 0;
        
        for (int move : moves) {
            removedDolls += processMove(lanes[move - 1], bucket);
        }
        
        return removedDolls;
    }
    
    private ArrayDeque<Integer>[] initializeLanes(int[][] board) {
        
        int size = board.length;
        ArrayDeque<Integer>[] lanes = new ArrayDeque[size];
        
        // 각 레인을 초기화
        for (int i = 0; i < size; i++) {
            lanes[i] = new ArrayDeque<>();
        }
        
        // 보드의 인형들을 레인에 추가
        for (int i = board.length - 1; i >= 0; i--) {
            for (int j = 0; j < board[i].length; j++) {
                if (board[i][j] > 0) {
                    lanes[j].push(board[i][j]);
                }
            }
        }
        
        return lanes;
    }
    
    private int processMove(ArrayDeque<Integer> lane, ArrayDeque<Integer> bucket) {
        // 레인이 비어있으면 처리하지 않음
        if (lane.isEmpty()) {
            return 0;
        }
        
        int doll = lane.pop();
        if (!bucket.isEmpty() && bucket.peek() == doll) {
            bucket.pop();
            return 2;  // 인형 2개가 제거됨
        } else {
            bucket.push(doll);
            return 0;
        }
    }
}